Run firewalld in CI

robmry opened the pull request:
**- What I did**

Found that CI's firewalld tests weren't using firewalld, so ...

### Add test util "FirewalldRunning"

### Firewalld should use its nftables backend

- introduced by https://github.com/moby/moby/pull/48638

Since commit cfc049c the dev container has been using iptables-nft, but the Dockerfile configured firewalld to use its iptables backend (if firewalld is enabled, which it hasn't been in CI).

The integration tests run ok for a while, then some things start failing with the less than obvious message:

`Error: COMMAND_FAILED: UNKNOWN_ERROR: nonexistent or underflow of priority count`

So, let firewalld use its default nftables backend.

### Firewalld: Skip tests that run dockerd in an L3Segment

The daemon runs in a separate netns, but when it wants to create an iptables rule it sends a dbus message to firewalld - which is running in the host's netns.

### TestIsolated for IPv6 is broken under firewalld

Skip the test for now, it's tracked at:
- https://github.com/moby/moby/issues/49680

### Skip flaky part of TestAccessPublishedPortFromHost

With firewalld enabled in CI, `TestAccessPublishedPortFromHost/userland-proxy=true/IPv6=true` consistently fails when trying to use a link-local address on eth0 (it's ok for the ULL added by the test).

In a local moby dev container, it passes - although it sometimes fails when making its request to the host's ::1.

Tracked at:
- https://github.com/moby/moby/issues/49695

### Firewalld: skip unit tests that run in their own netns

Firewalld runs in the host netns. So, tests running in their own netns can't check firewalld's iptables rules.

### Fix TestPassthrough

Doesn't look like it would ever have worked, but:
- init the dbus connection to avoid a segv
- include the chain name when creating the rule
- remove the test rule if it's created

### Install and run firewalld for CI's firewalld tests

Firewalld testing was ...
- initially added in https://github.com/moby/moby/pull/48603
- partially reverted in https://github.com/moby/moby/pull/48788
- re-added in https://github.com/moby/moby/pull/48756

The github action running bake expected FIREWALLD to be set, but DOCKER_FIREWALLD was set instead, so firewalld wasn't installed in the dev image.

The dind-systemd script expected DOCKER_FIREWALLD to be set if it needed to run firewalld, and it was. But it had no effect.

In CI, bake builds the image then make runs it - and they use the same env. So, align on FIREWALLD (as it's not a docker feature).

**- How I did it**

**- How to verify it**

**- Human readable description for the release notes**
```markdown changelog

```

robmry committed:
Install and run firewalld for CI's firewalld tests

The github action running bake expected FIREWALLD to be set, but
DOCKER_FIREWALLD was set instead, so firewalld wasn't installed
in the dev image.

The dind-systemd script expected DOCKER_FIREWALLD to be set if it
needed to run firewalld, and it was. But it had no effect.

In CI, bake builds the image then make runs it - and the use the
same env. So, align on FIREWALLD (as it's not a docker feature).

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Firewalld should use its nftables backend

Since commit cfc049c the dev container has been using iptables-nft,
but the Dockerfile configured firewalld to use its iptables backend
(if firewalld is enabled, which it accidentally hasn't been in CI).

The integration tests run ok for a while, then some things start
failing with the less than obvious message:

  Error: COMMAND_FAILED: UNKNOWN_ERROR: nonexistent or underflow of priority count

So, let firewalld use its default nftables backend.

Signed-off-by: Rob Murray <rob.murray@docker.com>

thaJeztah commented on api/types/system/info.go:
```diff
@@ -73,6 +73,7 @@ type Info struct {
 	SecurityOptions     []string
 	ProductLicense      string               `json:",omitempty"`
 	DefaultAddressPools []NetworkAddressPool `json:",omitempty"`
+	FirewallBackend     string               `json:",omitempty"`
```
Wondering how structured we'd want this info to be (more so if things are transitional); an alternative could be to have something similar to what we have for storage drivers, where the "driver-info" provides arbitrary data about the driver used; it's parseable, but we don't provide an exhaustive list (if any) for formal fields, as they're more "informative" and not strictly part of the API contract.

robmry replied:
At the moment it can be `iptables` or `iptables+firewalld`. Soon, it'll also be `nftables` or `firewalld` then eventually we'll get rid of iptables backend.

But, for firewalld, I'll also want to add a "last reload" time (to use in tests that trigger a reload and need to wait for it to complete).

So, are you thinking it should be something like this? ...

`"FirewallBackend": "backend=iptables+firewalld,firewalld_reload=<timestamp>"`

robmry replied:
As discussed - changed to `[][2]string`.

tianon replied:
If we're emulating a map, why not use a map? 😅

(This is an informational question, not a change suggestion, to be clear! ❤️)

robmry replied:
@thaJeztah wants a stable order for the fields.

But, the code is no longer in this PR ... I want to get the firewalld tests running, and move on. Will come back to the API change later.

robmry committed:
Firewalld: Skip tests that run dockerd in an L3Segment

The daemon runs in a separate netns, but when it wants to create
an iptables rule it sends a dbus message to firewalld - which is
running in the host's netns.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
TestIsolated for IPv6 is broken under firewalld

Skip the test for now.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Skip flaky part of TestAccessPublishedPortFromHost

With firewalld enabled in CI, TestAccessPublishedPortFromHost/userland-proxy=true/IPv6=true
consistently fails when trying to use a link-local address on
eth0 (it's ok for the ULL added by the test).

In a local moby dev container, it passes - although it sometimes
fails when making its request to the host's ::1.

Signed-off-by: Rob Murray <rob.murray@docker.com>

thaJeztah commented on api/swagger.yaml:
```diff
@@ -6856,6 +6856,13 @@ definitions:
               description: "The network pool size"
               type: "integer"
               example: "24"
+      FirewallBackend:
```
Oh! This would require a new API version, so we'd have to;

- update the API version (and default API version used)
- add a note to  `docs/api/version-history.md` https://github.com/moby/moby/blob/264fc2fac838e05deaa3c9d288fd93b4cb06f0d4/docs/api/version-history.md#v148-api-changes
- update the `/info` endpoint to omit the information for older API versions


I'd suggest doing the first step (updating the API version) in a separate PR for visibility, and so that we can link to it from our release notes; similar to https://github.com/moby/moby/commit/2b43979395c8d97f2c4b698b9cd3558101e881fd


(there's no need yet to add a `docs/api/v1.49.yaml` file; we can do so when we're about to do a release)

robmry replied:
Oh, yes, thank you.

So, once the API version bump is merged, we'll have to start cherry-picking to 28.x for patch releases (or make a minor release) ... that might be a bit unpopular, particularly for something as peripheral as this.

I could just drop the first commit from this PR - although I'll need to add it back in the next few weeks, as part of the nftables work. But it might avoid the cherry-picking for a bit longer.

Any preferences? @thaJeztah, @vvoland, @thompson-shaun

thaJeztah replied:
Yeah, overall, I don't think it should be a bit problem if we do a minor release next (and no need to use from a separate branch). That said, if there's no real user-facing "features" to be found, that may be slightly odd (but still technically correct for the API that is).

robmry replied:
Ok, with that, and all the extra comments ... I think the `FirewallBackend` field is more trouble than it's worth at the moment.

We need to get the firewalld tests running, that's the objective of this PR - the API change was only a way to prove it wasn't running, and in-prep for nftables.

So, I'll drop the API change for now, and come back to it as part of the nftables work.

robmry replied:
Done.

thaJeztah reviewed (changes requested):
api changes itself LGTM, but left a comment about updating the API version

thaJeztah commented on api/types/system/info.go:
```diff
@@ -73,6 +73,7 @@ type Info struct {
 	SecurityOptions     []string
 	ProductLicense      string               `json:",omitempty"`
 	DefaultAddressPools []NetworkAddressPool `json:",omitempty"`
+	FirewallBackend     [][2]string          `json:",omitempty"`
```
We've not been great at doing this, and relied on Golang doing "whatever golang chose", but we should probably start setting explicit names for `json` to indicate "this is what we want the name to be when marshalled);

```suggestion
	FirewallBackend     [][2]string          `json:"FirewallBackend,omitempty"`
```

robmry replied:
Dropped for now, but will-do when I get back to it.

thaJeztah commented on internal/testutils/networking/iptables.go:
```diff
@@ -44,3 +45,8 @@ func SetFilterForwardPolicies(t *testing.T, policy string) {
 	}
 }
+
+func FirewalldRunning() bool {
+	state, err := exec.Command("firewall-cmd", "--state").CombinedOutput()
```
As this is brand new; perhaps we could pass the context?

```suggestion
func FirewalldRunning(ctx context.Context) bool {
	state, err := exec.CommandContext(cox, "firewall-cmd", "--state").CombinedOutput()
```

thaJeztah commented on libnetwork/controller_linux.go:
```diff
@@ -12,6 +12,14 @@ import (
 	"github.com/docker/docker/libnetwork/osl"
 )
 
+// FirewallBackend returns the name of the firewall backend for "docker info".
+func (c *Controller) FirewallBackend() [][2]string {
+	if iptables.FirewalldRunning() {
```
And pass the context here;
```suggestion
// FirewallBackend returns the name of the firewall backend for "docker info".
func (c *Controller) FirewallBackend(ctx context.Context) [][2]string {
	if iptables.FirewalldRunning(cox) {
```

robmry committed:
Add test util "FirewalldRunning"

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented on api/swagger.yaml:
```diff
@@ -6856,6 +6856,13 @@ definitions:
+
+          Omitted on Windows hosts.
+        type: "string"
+        example: "iptables"
```
Thank you ... I've dropped the API change, but will come back to it.

thaJeztah replied:
If we (currently) only need it in tests, can we use an env-var that's set for those test suites, or won't that work?

robmry replied:
I added a test that really just logged whether the daemon thought firewalld was running - because it was so surprising that firewalld wasn't running for the firewalld CI tests that I wanted to prove it.

So, an env var wouldn't help - I wanted to ask the daemon what firewall-backend it was using, and putting that in the `Info` response will definitely be useful once nftables is plumbed in. But, for now, dropping it is not a problem.

robmry replied:
(I added the test-util version of `FirewalldRunning` later - as a way to skip tests. That's all we need for now.)

robmry committed:
Firewalld: skip unit tests that run in their own netns

Firewalld runs in the host netns. So, tests running in
their own netns can't check firewalld's iptables rules.

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry committed:
Fix TestPassthrough

Doesn't look like it would ever have worked, but:
- init the dbus connection to avoid a segv
- include the chain name when creating the rule
- remove the test rule if it's created

Signed-off-by: Rob Murray <rob.murray@docker.com>

robmry commented:
Rebased on https://github.com/moby/moby/pull/49671 and fixed more test failures.

thaJeztah reviewed (approved):
LGTM

akerouanton commented on Makefile:
```diff
@@ -38,7 +38,7 @@ DOCKER_ENVS := \
 	-e DOCKER_DEBUG \
 	-e DOCKER_EXPERIMENTAL \
-	-e DOCKER_FIREWALLD \
+	-e FIREWALLD \
```
This list is sorted alphabetically. Can you put it after `DELVE_PORT` please?

akerouanton reviewed (approved):
Small nit, otherwise LGTM.

chandrakanthbala commented:
Hi @robmry @thaJeztah @akerouanton 

We recently migrated to RHEL 8.10 and are using Docker CE 27.4.0. We are encountering the following error, as described in this issue.

```Error: COMMAND_FAILED: UNKNOWN_ERROR: nonexistent or underflow of priority count```

As of now, we haven't made any changes to firewalld; we're using the default out-of-the-box configuration. Could you please let me know what changes are required to resolve this issue on the RHEL 8.10 server? Does any recent version of Docker fix this automatically, or do we still need to make changes to firewalld?

robmry commented:
> We recently migrated to RHEL 8.10 and are using Docker CE 27.4.0. We are encountering the following error, as described in this issue.

Hi @chandrakanthbala - the issues fixed here relate to Docker's test environment, they're not product changes.

What caused the problem for us was we were using `iptables-nft` (so, `iptables --version` would show something like `iptables v1.8.9 (nf_tables)`) ... but were configuring firewalld to expect the legacy "xtables" backend - compatible with `iptables-legacy` (`iptables v1.8.9 (legacy)`).

In firewalld's config, we needed `FirewallBackend=nftables`, because `FirewallBackend=iptables` gave us the error.

So, you could try checking those things match on your host.

chandrakanthbala commented:
> `FirewallBackend`

We are using ```FirewallBackend=nftables``` and ```iptables --version``` gives ```iptables v1.8.5 (nf_tables)```. Could you give any insights into what might be causing the error?"

thaJeztah commented:
Quick search shows the error looks to be produced in one of these

https://github.com/firewalld/firewalld/blob/ad6f103e1d656d263116ef0178c08792e32aac84/src/firewall/core/ipXtables.py#L303-L305

https://github.com/firewalld/firewalld/blob/ad6f103e1d656d263116ef0178c08792e32aac84/src/firewall/core/nftables.py#L362-L364
